pkg/controller/istiocsr/certificates.go

```diff
@@ -0,0 +1,144 @@
+	}
+
+	dnsNames := make([]string, len(istiocsr.Spec.IstioCSRConfig.IstiodTLSConfig.CertificateDNSNames))
+	copy(dnsNames, istiocsr.Spec.IstioCSRConfig.IstiodTLSConfig.CertificateDNSNames)
+
+	// Also need to add a DNS SAN for every requested revision, except for default revision, which will not be
+	// included in the DNS name.
+	for _, revision := range istiocsr.Spec.IstioCSRConfig.Istio.Revisions {
+		if revision == "" {
+			continue
+		}
+		if revision == "default" {
+			name := fmt.Sprintf(istiodCertificateDefaultDNSName, istiocsr.Spec.IstioCSRConfig.Istio.Namespace)
+			dnsNames = append(dnsNames, name)
+			continue
+		}
+		name := fmt.Sprintf(istiodCertificateRevisionBasedDNSName, revision, istiocsr.Spec.IstioCSRConfig.Istio.Namespace)
+		dnsNames = append(dnsNames, name)
+	}
+	certificate.Spec.DNSNames = dnsNames
+
+	certificate.Spec.URIs = []string{
+		fmt.Sprintf(istiodCertificateSpiffeURIFmt, istiocsr.Spec.IstioCSRConfig.IstiodTLSConfig.TrustDomain, istiocsr.Spec.IstioCSRConfig.Istio.Namespace),
+	}
+
+	certificate.Spec.Duration = istiocsr.Spec.IstioCSRConfig.IstiodTLSConfig.CertificateDuration
+	if certificate.Spec.Duration == nil {
+		certificate.Spec.Duration = &metav1.Duration{Duration: DefaultCertificateDuration}
+	}
+
+	certificate.Spec.RenewBefore = istiocsr.Spec.IstioCSRConfig.IstiodTLSConfig.CertificateRenewBefore
+	if certificate.Spec.RenewBefore == nil {
+		certificate.Spec.RenewBefore = &metav1.Duration{Duration: DefaultCertificateRenewBeforeDuration}
+	}
+
+	certificate.Spec.PrivateKey.Algorithm = certmanagerv1.PrivateKeyAlgorithm(istiocsr.Spec.IstioCSRConfig.IstiodTLSConfig.SignatureAlgorithm)
+	if certificate.Spec.PrivateKey.Algorithm == "" {
+		certificate.Spec.PrivateKey.Algorithm = DefaultSignatureAlgorithm
+	}
+
+	certificate.Spec.PrivateKey.Size = istiocsr.Spec.IstioCSRConfig.IstiodTLSConfig.PrivateKeySize
+	if certificate.Spec.PrivateKey.Size == 0 {
+		if certificate.Spec.PrivateKey.Algorithm == certmanagerv1.RSAKeyAlgorithm {
+			certificate.Spec.PrivateKey.Size = DefaultRSAPrivateKeySize
+		}
+		if certificate.Spec.PrivateKey.Algorithm == certmanagerv1.ECDSAKeyAlgorithm {
+			certificate.Spec.PrivateKey.Size = DefaultECDSA384PrivateKeySize
+		}
+	}
+	if (certificate.Spec.PrivateKey.Algorithm == certmanagerv1.RSAKeyAlgorithm && certificate.Spec.PrivateKey.Size < DefaultRSAPrivateKeySize) ||
+		(certificate.Spec.PrivateKey.Algorithm == certmanagerv1.ECDSAKeyAlgorithm && certificate.Spec.PrivateKey.Size != DefaultECDSA256PrivateKeySize && certificate.Spec.PrivateKey.Size != DefaultECDSA384PrivateKeySize) {
+		return fmt.Errorf("certificate parameters PrivateKeySize and SignatureAlgorithm do not comply")
+	}
+
+	certificate.Spec.IssuerRef = certmanagermetav1.ObjectReference{
+		Kind:  istiocsr.Spec.IstioCSRConfig.CertManager.IssuerRef.Kind,
+		Group: istiocsr.Spec.IstioCSRConfig.CertManager.IssuerRef.Group,
+		Name:  istiocsr.Spec.IstioCSRConfig.CertManager.IssuerRef.Name,
+	}
```

Comment:
need to check here before to ensure `istiocsr.Spec.IstioCSRConfig.CertManager` is not nil.